crates/hir-ty/src/diagnostics/match_check/pat_analysis.rs

```diff
@@ -89,9 +115,18 @@ impl<'p> MatchCheckCtx<'p> {
         }
     }
 
-    fn variant_id_for_adt(ctor: &Constructor<Self>, adt: hir_def::AdtId) -> Option<VariantId> {
+    fn variant_id_for_adt(
+        db: &'p dyn HirDatabase,
+        ctor: &Constructor<Self>,
+        adt: hir_def::AdtId,
+    ) -> Option<VariantId> {
         match ctor {
-            &Variant(id) => Some(id.into()),
+            Variant(id) => {
+                let hir_def::AdtId::EnumId(eid) = adt else {
+                    panic!("bad constructor {ctor:?} for adt {adt:?}")
```

Comment:
Maybe `stdx::never` and `return None` here so we don't panic?

Reply:
If we return `None` we panic only a tiny bit later because most of the calls of this function `unwrap()` the output

Reply:
Want me to change that anyway?

Reply:
panicking early seems better then